In [1]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat

mnist = loadmat('datasets/mnist-original.mat')
X = mnist['data']
y = mnist['label']
X = X.T
y = y.reshape(70000,)

In [2]:
from sklearn.decomposition import PCA
pca = PCA(0.90)
pca.fit(X)
X_reduction = pca.transform(X)

In [3]:
pca.n_components_

87

In [4]:
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(n_estimators=1000, max_leaf_nodes=5000, oob_score=True, random_state=666, n_jobs=-1)
rf_clf.fit(X_reduction, y)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=5000,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=-1,
            oob_score=True, random_state=666, verbose=0, warm_start=False)

In [5]:
rf_clf.oob_score_

0.9547

### n_estimators=1000, max_leaf_nodes=5000的意思是 使用了1000个决策树的分类器，而每一个分类器最多只能有5000个叶子节点

### 这段代码中，使用的取数据的方法是老师课上介绍的自助法，oob_score_的结果0.9547是对从来没有取到过的数据进行预测所得到的准确率

### 代码中使用的测试数据集是mnist手写数据集，其官方网站是http://yann.lecun.com/exdb/mnist/

In [7]:
from PIL import Image

In [46]:
three = Image.open('three.png' ).convert('L')
three = np.array(three)

In [47]:
three.shape

(28, 28)

In [32]:
three = three.reshape(-1,)

In [34]:
three = 255 - three

In [35]:
three

array([  2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
         2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
         2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
         2,   2,  11,  18,  15,   2,   2,   2,   2,   2,   2,   2,   2,
         2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
         2,  17, 101, 191, 255, 255, 255, 211,  80,   3,   2,   2,   2,
         2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
         2,  19, 136, 244, 255, 255, 255, 255, 255, 255, 255, 167,   4,
         2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
         2,   2,   2, 211, 255, 255, 255, 255, 253, 241, 251, 255, 255,
       255, 116,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
         2,   2,   2,   2,   4, 244, 255, 255, 212, 113,  27,   2,  30,
       162, 255, 255, 251,  45,   2,   2,   2,   2,   2,   2,   2,   2,
         2,   2,   2,   2,   2,   2,   2,  73, 160,  84,   3,   

In [42]:
three_reduction = pca.transform(three.reshape(1,-1))
rf_clf.predict(three_reduction)

array([3.])

### 可以看出，代码成功预测出了3